### PR TITLE
First part of circuitry update.

### PR DESCRIPTION
**Pull request by arsserpentarium**
1.Fixed pathfinder circuit …
2.Hypo injector and pump now can be inversed if you nive negative amount.
3.Hypo injector and smoke generator has output pin with amount of stored liquid now.Also they have pulse out pins ,which will be activated after their work.
4.All reagent containers are acidproof now and have pin with self reference(to make it more easy to link them)
5.advanced locator. Like adjastent, but with adjustable up to 8 tiles radius and possibility to find items by name and desctiption.
6.drone capacity and complexity was raised.
7.Now you can wrench large assembly into place.
8.Adjastent locator have 2 pulse out pins. One will be activated if it find something, And one if doesn't
9.reagent scaner. It's like a beaker. But can give a list of contained reagents.
10.reagent filter. Pump,that can transfer only desired reagents,or all,except unwanted.
11.examiner. returns name,description ,relative coordiinates,distance,amount of reagents
12. Internal and external battery sensers. A lot like power monitor in transmission circuit. But without transmission. First returns state of assembly's own battety. Second returns state of any battery in view.
13.encoder/decoder of refs.To send refs throught EPV2
14.strings: find element in string. explode string by delimeter.

### CODE REVIEW

File: code/modules/integrated_electronics/subtypes/converters.dm [hunk 1]
```diff
@@ -67,6 +67,40 @@
...
+	if(A && istype(A))
+		result = "\ref[A]"
+
+	set_pin_data(IC_OUTPUT, 1, result)
+	push_data()
+	activate_pin(2)
+
+/obj/item/integrated_circuit/converter/refdecode
+	name = "reference decoder"
+	desc = "This circuit can convert a encoded reference to actual reference."
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/converters.dm]: an encoded

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 2]
```diff
@@ -104,10 +106,54 @@
 
 	push_data()
 	activate_pin(2)
+/*
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: Why are you adding a massive block of commented out code?

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 3]
```diff
@@ -130,7 +176,7 @@
 	var/mob/living/carbon/human/H = get_pin_data_as_type(IC_INPUT, 1, /mob/living/carbon/human)
 	if(!istype(H)) //Invalid input
 		return
-	if(H.Adjacent(get_turf(src))) // Like normal analysers, it can't be used at range.
+	if(H in view(get_turf(H))) // Like medbot's analyzer it can be used in range..
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: Why was this changed, precisely?

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 4]
```diff
@@ -145,6 +191,53 @@
 	push_data()
 	activate_pin(2)
 
+/obj/item/integrated_circuit/input/examiner
+	name = "examiner"
+	desc = "It' s a little machine vision system. It can return the name, description, distance, \
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: It's, no space.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 5]
```diff
@@ -197,7 +291,115 @@
 		valid_things.Add(thing)
 	if(valid_things.len)
 		O.data = weakref(pick(valid_things))
+		activate_pin(2)
+	else
+		activate_pin(3)
 	O.push_data()
+/*
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: Again, why the massive block of commented code?

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 6]
```diff
@@ -197,7 +291,115 @@
...
+        activate_pin(2)
+    else
+        O.push_data()
+        activate_pin(3)
+*/
+/obj/item/integrated_circuit/input/advanced_locator
+	complexity = 6
+	name = "advanced locator"
+	desc = "This is needed for certain devices that demand a reference for a target to act upon. This type locates something \
+	that is standing in given radius up to 8 meters"
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: of up to 8 meters.

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 7]
```diff
@@ -197,7 +291,115 @@
...
+	extended_desc = "The first pin requires a ref to a kind of object that you want the locator to acquire. If  This means that it will \
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: You don't need that If

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 8]
```diff
@@ -197,7 +291,115 @@
...
+	give refs to nearby objects that are similar. If this pin is string, locator will search\
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: Similar to what?
If this pin is a string, the locator

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 9]
```diff
@@ -197,7 +291,115 @@
...
+	 item by matching desired text in name + description. If more than one valid object is found nearby, it will choose one of them at \
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: search item means what?

File: code/modules/integrated_electronics/subtypes/logic.dm [hunk 10]
```diff
@@ -55,6 +55,98 @@
...
+	Q.data = !lstate ? TRUE : FALSE
+	if(get_pin_data(IC_OUTPUT, 1))
+		activate_pin(2)
+	else
+		activate_pin(3)
+	push_data()
+
+/obj/item/integrated_circuit/logic/binary/gdlatch
+	name = "gated D latch"
+	desc = "This gate is synchronysed gated D latch."
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/logic.dm]: is a

File: code/modules/integrated_electronics/subtypes/logic.dm [hunk 11]
```diff
@@ -55,6 +55,98 @@
...
+	Q.data = !lstate ? TRUE : FALSE
+	if(get_pin_data(IC_OUTPUT, 1))
+		activate_pin(2)
+	else
+		activate_pin(3)
+	push_data()
+
+/obj/item/integrated_circuit/logic/binary/rslatch
+	name = "RS latch"
+	desc = "This gate is synchronysed RS latch. If both R and S are true, state will not change."
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/logic.dm]: is a
the state

File: code/modules/integrated_electronics/subtypes/logic.dm [hunk 12]
```diff
@@ -55,6 +55,98 @@
 /obj/item/integrated_circuit/logic/binary/equals/do_compare(var/datum/integrated_io/A, var/datum/integrated_io/B)
 	return A.data == B.data
 
+/obj/item/integrated_circuit/logic/binary/jklatch
+	name = "JK latch"
+	desc = "This gate is synchronysed JK latch."
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/logic.dm]: is a

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 13]
```diff
@@ -431,3 +633,83 @@
...
+
+			set_pin_data(IC_OUTPUT, 1, assembly.battery.charge)
+			set_pin_data(IC_OUTPUT, 2, assembly.battery.maxcharge)
+			set_pin_data(IC_OUTPUT, 3, 100*assembly.battery.charge/assembly.battery.maxcharge)
+	push_data()
+	activate_pin(2)
+
+/obj/item/integrated_circuit/input/externalbm
+	name = "external battery monitor"
+	desc = "This can help to watch battery state of any device in view"
```
Comment (Anewbe) [inline code/modules/integrated_electronics/subtypes/input.dm]: "This can help watch the battery level of any device in range."

File: code/modules/integrated_electronics/subtypes/input.dm (hunk 3 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/input.dm]: This was changed to allow remote work of health sensor. I think, that it should be posible to recreate functions of any bot(farmbot,medbot,securitron, cleanbot,floorbot) with circuitry arsenal. For means of logic and more freedom of creativity.
Medbot can read this params in line of sight.So electrical bot should too.

File: code/modules/integrated_electronics/subtypes/input.dm (hunk 13 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/input.dm]: I changed, but this device can read data only in line of sight.

**Comment by arsserpentarium** [comment]
All isues fixed

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 14]
```diff
@@ -105,9 +107,11 @@
 	push_data()
 	activate_pin(2)
 
+
+
 /obj/item/integrated_circuit/input/adv_med_scanner
 	name = "integrated advanced medical analyser"
-	desc = "A very small version of the common medical analyser.  This allows the machine to know how healthy someone is.  \
+	desc = "A very small version of the medbot's medical analyser.  This allows the machine to know how healthy someone is.  \
```
Comment (Belsima) [inline code/modules/integrated_electronics/subtypes/input.dm]: Why was this changed to specify medbot?

Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/input.dm]: but medbot can remotely scan your health.
I don't know, where is logic, but there is a fact: if you want to recreate medbot's functions with circuitry, you'll need remote health scan.
